### PR TITLE
top: Replace upip with mip everywhere.

### PR DESCRIPTION
**Pull request by jimmo**
Updates all README.md and docs, and manifests to `require("mip")`.

Also extend and improve the documentation on freezing and packaging, in particular a fairly significant overhaul of `manifest.rst` and `packaging.rst`.

_This work was funded through GitHub Sponsors._

### CODE REVIEW

**Comment by jimmo** [comment]
See https://github.com/micropython/micropython-lib/pull/542 for more info about `mip`.

File: docs/develop/gettingstarted.rst [hunk 1]
```diff
@@ -322,7 +322,7 @@ tests
 
 tools
 
-  Contains helper tools including the ``upip`` and the ``pyboard.py`` module.
+  Contains scripts used by the build and CI process, as well as user tools such as ``pyboard.py``.
```
Comment (dpgeorge) [inline docs/develop/gettingstarted.rst:325]: maybe add "and mpremote"?

File: docs/reference/glossary.rst [hunk 2]
```diff
@@ -151,16 +151,24 @@ Glossary
         machine-independent features. It can also function in a similar way to
         :term:`CPython`'s ``python`` executable.
 
+    mip
+        A package installer for MicroPython. This can be used on-device on
+        network-capable boards and internally by tools such as :term:`mpremote`
+        to install packages either from :term:`micropython-lib` or GitHub.
```
Comment (dpgeorge) [inline docs/reference/glossary.rst:157]: Suggested change:
```
A package installer for MicroPython (mip - mip installs packages). It installs
Python packages packages either from :term:`micropython-lib` or GitHub,
or arbitrary URLs.  mip can be used on-device on
network-capable boards, and internally by tools such as :term:`mpremote`.  
```

File: docs/reference/manifest.rst [hunk 3]
```diff
@@ -1,35 +1,176 @@
...
+
+Custom manifest files should not live in the main MicroPython repository. You
+should keep them in version control with the rest of your project.
+
+Typically a manifest used for compiling firmware will need to include the port
+manifest, which might include frozen modules that are required for the board to
+function. If you just want to add additional modules to an existing board, then
+include the board manifest (which will in turn include the port manifest).
+
+The ``opt`` kwarg can be set on the various functions, this controls the
```
Comment (dpgeorge) [inline docs/reference/manifest.rst:62]: kwarg -> keyword argument

also, maybe this paragraph makes more sense down below, closer to the functions themselves?

File: docs/reference/packages.rst [hunk 4]
```diff
@@ -1,314 +1,150 @@
...
+==================
+
+Installing packages with ``mip``
+--------------------------------
+
+Network-capable boards include the ``mip`` module, which can install packages
+from :term:`micropython-lib` and from third-party sites (including GitHub).
+
+``mip`` ("mip installs packages") is similar in concept to Python's ``pip`` tool,
+however it does not use the PyPi index, rather it uses :term:`micropython-lib`
```
Comment (dpgeorge) [inline docs/reference/packages.rst:13]: PyPi -> PyPI

File: docs/reference/packages.rst [hunk 5]
```diff
@@ -1,314 +1,150 @@
...
+    >>> mip.install("github:org/repo", version="branch-or-tag")
+
+
+Using ``mip`` on the Unix port
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+On the Unix port, ``mip`` can be used at the REPL as above, and also by using ``-m``::
+
+    $ ./micropython -m mip install pkgname-or-url
+    $ ./micropython -m mip install pkgname-or-url@version
```
Comment (dpgeorge) [inline docs/reference/packages.rst:66]: this functionality doesn't exist yet, right?

File: docs/reference/packages.rst [hunk 6]
```diff
@@ -1,314 +1,150 @@
...
+The ``--install-target=path`` and ``--no-install-mpy`` arguments can be set::
+
+    $ ./micropython -m mip install --install-target=third-party pkgname
+    $ ./micropython -m mip install --install-target=third-party-source --no-install-mpy pkgname
+
+Installing packages with ``mpremote``
+-------------------------------------
+
+The :term:`mpremote` tool also includes the same functionality as ``mip`` and can
+be used to install packages to a connected device::
```
Comment (dpgeorge) [inline docs/reference/packages.rst:77]: maybe make it clear that mpremote runs on the host and installs on the target, and so the target doesn't need to have networking support

File: docs/reference/packages.rst [hunk 7]
```diff
@@ -1,314 +1,150 @@
...
+        ["collections-defaultdict", "latest"],
+        ["os-path", "latest"]
+      ],
+      "version": "0.2"
+    }
+
+This includes two files, hosted at a GitHub repo named
+``org/micropython-mlx90640``, which install into the ``mlx90640`` directory on
+the device. It depends on ``collections-defaultdict`` and ``os-path`` which will
+be installed automatically.
```
Comment (dpgeorge) [inline docs/reference/packages.rst:127]: is it worth adding some docs about creating and using your own index?  maybe in a new section?

**Comment by dpgeorge** [comment]
TODO:
- [ ] add removed upip to micropython-lib

File: docs/reference/packages.rst (hunk 5 above)
Comment (jimmo) [reply docs/reference/packages.rst:66]: No, but PR incoming.

File: docs/reference/packages.rst (hunk 7 above)
Comment (jimmo) [reply docs/reference/packages.rst:127]: Separate PR :)

**Comment by jimmo** [comment]
Addressed all comments.

**Comment by jimmo** [comment]
Added a fix to `tools/manifestfile.py` to fix Python 2 compatibility.

File: docs/reference/packages.rst (hunk 7 above)
Comment (dpgeorge) [reply docs/reference/packages.rst:130]: OK!

File: docs/reference/glossary.rst [hunk 8]
```diff
@@ -151,16 +151,26 @@ Glossary
         machine-independent features. It can also function in a similar way to
         :term:`CPython`'s ``python`` executable.
 
+    mip
+        A package installer for MicroPython (mip - "mip installs packages"). It
+        installs Python packages packages either from :term:`micropython-lib`
```
Comment (dpgeorge) [inline docs/reference/glossary.rst:156]: "packages packages"

**Pull request merged**